File: ios/ClawChat/Sources/ClawChat/Wire.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// Wire format — mirrors plugin/src/wire.ts byte-for-byte.
//
// One JSON object per UDP datagram. Multi-chunk reassembly keyed by `id`
// with (`seq`, `total`).

import Foundation
import CryptoKit

public enum Wire {

    public static let version: Int = 1
    public static let maxEnvelopeBytes: Int = 1024

    public enum Kind: String, Codable {
        case user
        case agent
        case media
        case ack
        case error
    }

    public enum MediaKind: String, Codable, Equatable {
        case image
        case file
        case audio
    }

    public enum MediaFrom: String, Codable, Equatable {
        case user
        case agent
    }

    public struct Envelope: Codable, Equatable {
        public var v: Int
        public var kind: Kind
        public var id: String
        public var ts: Int64

        // Text fields (kind ∈ {user, agent} or media caption)
        public var text: String?
        public var seq: Int?
        public var total: Int?

        // Error fields
        public var code: String?

        // Media fields
        public var from: MediaFrom?
        public var media: MediaKind?
        public var data: String?       // base64 payload
        public var filename: String?
        public var mime: String?
        public var totalBytes: Int?
        public var caption: String?

        // HMAC-SHA256 over the canonicalized envelope (no hmac field), base64.
        public var hmac: String?

        public init(
            v: Int = Wire.version,
            kind: Kind,
            id: String,
            ts: Int64,
            text: String? = nil,
            seq: Int? = nil,
            total: Int? = nil,
            code: String? = nil,
            from: MediaFrom? = nil,
            media: MediaKind? = nil,
            data: String? = nil,
            filename: String? = nil,
            mime: String? = nil,
            totalBytes: Int? = nil,
            caption: String? = nil,
            hmac: String? = nil
        ) {
            self.v = v
            self.kind = kind
            self.id = id
            self.ts = ts
            self.text = text
            self.seq = seq
            self.total = total
            self.code = code
            self.from = from
            self.media = media
            self.data = data
            self.filename = filename
            self.mime = mime
            self.totalBytes = totalBytes
            self.caption = caption
            self.hmac = hmac
        }
    }

    // MARK: - HMAC (shared-secret authorization)

    /// Canonicalize an envelope for HMAC — strip the `hmac` field, sort keys,
    /// serialize as compact JSON. MUST match the plugin's canonicalForHmac.
    public static func canonicalForHmac(_ env: Envelope) throws -> Data {
        var dict: [String: Any] = [:]
        dict["v"] = env.v
        dict["kind"] = env.kind.rawValue
        dict["id"] = env.id
        dict["ts"] = NSNumber(value: env.ts)
        if let v = env.text { dict["text"] = v }
        if let v = env.seq { dict["seq"] = v }
        if let v = env.total { dict["total"] = v }
        if let v = env.code { dict["code"] = v }
        if let v = env.from { dict["from"] = v.rawValue }
        if let v = env.media { dict["media"] = v.rawValue }
        if let v = env.data { dict["data"] = v }
        if let v = env.filename { dict["filename"] = v }
        if let v = env.mime { dict["mime"] = v }
        if let v = env.totalBytes { dict["totalBytes"] = v }
        if let v = env.caption { dict["caption"] = v }
        // Sort keys; JSONSerialization with .sortedKeys gives us this.
        return try JSONSerialization.data(
            withJSONObject: dict,
            options: [.sortedKeys, .withoutEscapingSlashes]
        )
    }

    public static func sign(_ env: Envelope, secret: String) throws -> String {
        let canonical = try canonicalForHmac(env)
        let key = SymmetricKey(data: Data(secret.utf8))
        let mac = HMAC<SHA256>.authenticationCode(for: canonical, using: key)
        return Data(mac).base64EncodedString()
    }

    public static func verify(_ env: Envelope, secret: String) throws -> Bool {
        guard let provided = env.hmac, let providedBytes = Data(base64Encoded: provided) else {
            return false
        }
        let canonical = try canonicalForHmac(env)
        let key = SymmetricKey(data: Data(secret.utf8))
        let mac = HMAC<SHA256>.authenticationCode(for: canonical, using: key)
        return HMAC<SHA256>.isValidAuthenticationCode(
            mac,
            authenticating: canonical,
            using: key
        ) && providedBytes.elementsEqual(Data(mac))
    }

    public enum WireError: Error, CustomStringConvertible, Equatable {
        case tooLarge(bytes: Int)
        case notJSON(String)
        case schemaFailed(String)

        public var description: String {
            switch self {
            case .tooLarge(let n):       return "envelope \(n)B exceeds maxEnvelopeBytes (\(Wire.maxEnvelopeBytes))"
            case .notJSON(let m):        return "envelope not JSON: \(m)"
            case .schemaFailed(let m):   return "envelope failed schema check: \(m.prefix(200))"
            }
        }
    }

    public static func encode(_ env: Envelope) throws -> Data {
        let enc = JSONEncoder()
        // Match the JS encoder's compact / no-extra-keys output.
        enc.outputFormatting = []
        // Match TS field order: v, kind, id, ts, text, seq, total, code.
        // JSONEncoder ordering isn't guaranteed but stays stable enough for
        // small envelopes; the reassembler doesn't care about field order.
        let data = try enc.encode(env)
        if data.count > maxEnvelopeBytes {
            throw WireError.tooLarge(bytes: data.count)
        }
        return data
    }

    public static func decode(_ data: Data) throws -> Envelope {
        let dec = JSONDecoder()
        let env: Envelope
        do {
            env = try dec.decode(Envelope.self, from: data)
        } catch {
            throw WireError.notJSON(error.localizedDescription)
        }
        if env.v != version {
            throw WireError.schemaFailed("unsupported version \(env.v)")
        }
        if env.id.isEmpty {
            throw WireError.schemaFailed("missing id")
        }
        switch env.kind {
        case .user, .agent:
            if env.text == nil {
                throw WireError.schemaFailed("\(env.kind.rawValue) requires text")
            }
        case .media:
            if env.data == nil || env.seq == nil || env.total == nil
                || env.from == nil || env.media == nil {
                throw WireError.schemaFailed("media requires data + seq + total + from + media")
            }
        case .ack:
            break
        case .error:
            if env.code == nil || env.text == nil {
                throw WireError.schemaFailed("error requires code + text")
            }
        }
        return env
    }

    /// 24-char base36 id matching newId() in wire.ts.
    public static func newId() -> String {
        let t = String(Int64(Date().timeIntervalSince1970 * 1000), radix: 36)
            .padding(toLength: 9, withPad: "0", startingAt: 0)
        var rand = UInt64.random(in: 0..<UInt64(pow(Double(36), 12)))
        var chars: [Character] = []
        for _ in 0..<12 {
            let digit = Int(rand % 36)
            rand /= 36
            chars.append(base36(digit))
        }
        return t + String(chars.reversed())
    }

    private static func base36(_ d: Int) -> Character {
        if d < 10 { return Character(String(d)) }
        return Character(UnicodeScalar(97 + d - 10)!) // 'a'..'z'
    }

    /// Split a long text into N envelopes that each fit `maxEnvelopeBytes`.
    public static func chunk(
        text: String,
        kind: Kind,
        id: String = newId(),
        ts: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) -> [Envelope] {
        precondition(kind == .user || kind == .agent, "chunk only supports user/agent")
        let unsplit = Envelope(kind: kind, id: id, ts: ts, text: text)
        if let single = try? encode(unsplit), single.count <= maxEnvelopeBytes {
            return [unsplit]
        }

        // Estimate per-envelope overhead (with seq/total fields).
        let header = Envelope(kind: kind, id: id, ts: ts, text: "", seq: 1, total: 99)
        let overhead = (try? encode(header).count) ?? 200
        let room = max(64, maxEnvelopeBytes - overhead - 8)

        var pieces: [String] = []
        var idx = text.startIndex
        while idx < text.endIndex {
            // Binary search a piece length (in characters) that fits in `room` UTF-8 bytes.
            var lo = 1
            var hi = text.distance(from: idx, to: text.endIndex)
            var best = 1
            while lo <= hi {
                let mid = (lo + hi) / 2
                let end = text.index(idx, offsetBy: mid)
                let piece = String(text[idx..<end])
                if piece.utf8.count <= room {
                    best = mid
                    lo = mid + 1
                } else {
                    hi = mid - 1
                }
            }
            let endIdx = text.index(idx, offsetBy: best)
            pieces.append(String(text[idx..<endIdx]))
            idx = endIdx
        }

        return pieces.enumerated().map { (i, piece) in
            Envelope(
                kind: kind,
                id: id,
                ts: ts,
                text: piece,
                seq: i + 1,
                total: pieces.count
            )
        }
    }

    /// In-memory reassembler. Drop reassembly state for `id`s older than `maxAgeMs`.
    public final class Reassembler {
        private struct State {
            var received: [Int: Envelope]
            var total: Int
            var firstTs: Int64
        }
        private var parts: [String: State] = [:]
        private let kind: Kind

        public init(kind: Kind) {
            precondition(kind == .user || kind == .agent)
            self.kind = kind
        }

        public func gc(now: Int64 = Int64(Date().timeIntervalSince1970 * 1000), maxAgeMs: Int64 = 60_000) {
            for (id, state) in parts where now - state.firstTs > maxAgeMs {
                parts.removeValue(forKey: id)
            }
        }

        /// Feed one envelope. Returns the fully reassembled envelope if complete, else nil.
        public func push(_ env: Envelope) -> Envelope? {
            guard env.kind == kind else { return nil }
            guard let total = env.total, total > 1 else {
                // Single-chunk or unchunked → return as-is.
                if env.total == 1 || env.total == nil {
                    return env
                }
                return nil
            }
            guard let seq = env.seq, seq >= 1, seq <= total else { return nil }
            var st = parts[env.id] ?? State(received: [:], total: total, firstTs: env.ts)
            if st.total != total { return nil }
            st.received[seq] = env
            parts[env.id] = st
            if st.received.count < total { return nil }
            var combined = ""
            for i in 1...total {
                guard let part = st.received[i]?.text else { return nil }
                combined += part
            }
            parts.removeValue(forKey: env.id)
            return Envelope(kind: env.kind, id: env.id, ts: env.ts, text: combined)
        }
    }

    // MARK: - Media

    public struct MediaChunkInput {
        public let from: MediaFrom
        public let media: MediaKind
        public let bytes: Data
        public var filename: String?
        public var mime: String?
        public var caption: String?
        public var id: String?
        public var ts: Int64?

        public init(
            from: MediaFrom,
            media: MediaKind,
            bytes: Data,
            filename: String? = nil,
            mime: String? = nil,
            caption: String? = nil,
            id: String? = nil,
            ts: Int64? = nil
        ) {
            self.from = from
            self.media = media
            self.bytes = bytes
            self.filename = filename
            self.mime = mime
            self.caption = caption
            self.id = id
            self.ts = ts
        }
    }

    public struct ReassembledMedia: Equatable {
        public let id: String
        public let ts: Int64
        public let from: MediaFrom
        public let media: MediaKind
        public let bytes: Data
        public let filename: String?
        public let mime: String?
        public let caption: String?
    }

    /// Split a binary payload into media envelopes that fit MAX_ENVELOPE_BYTES.
    public static func chunkMedia(_ input: MediaChunkInput) -> [Envelope] {
        let id = input.id ?? newId()
        let ts = input.ts ?? Int64(Date().timeIntervalSince1970 * 1000)
        let totalBytes = input.bytes.count

        // Overhead estimates with/without metadata.
        let firstSample = Envelope(
            kind: .media,
            id: id,
            ts: ts,
            seq: 1,
            total: 999,
            from: input.from,
            media: input.media,
            data: "",
            filename: input.filename,
            mime: input.mime,
            totalBytes: totalBytes,
            caption: input.caption
        )
        let restSample = Envelope(
            kind: .media,
            id: id,
            ts: ts,
            seq: 999,
            total: 999,
            from: input.from,
            media: input.media,
            data: ""
        )
        // The encoded `restSample` has an empty `data: ""`. Once we put real
        // base64 in, the total envelope size = overhead + base64Length(N) -
        // 0 (we already counted the empty string + quotes). Conservative
        // safety margin to account for differences in JSON key ordering /
        // escaping across encoders.
        let firstOverhead = (try? encode(firstSample).count) ?? 300
        let restOverhead = (try? encode(restSample).count) ?? 200
        let firstDataRoom = max(64, maxEnvelopeBytes - firstOverhead - 32)
        let restDataRoom = max(64, maxEnvelopeBytes - restOverhead - 32)
        let firstSrcBytes = max(3, (firstDataRoom / 4) * 3)
        let restSrcBytes = max(3, (restDataRoom / 4) * 3)

        var chunks: [Data] = []
        var offset = 0
        let first = input.bytes.subdata(in: 0..<min(firstSrcBytes, totalBytes))
        chunks.append(first)
        offset += first.count
        while offset < totalBytes {
            let end = min(offset + restSrcBytes, totalBytes)
            chunks.append(input.bytes.subdata(in: offset..<end))
            offset = end
        }

        let total = chunks.count
        return chunks.enumerated().map { (idx, chunk) in
            var env = Envelope(
                kind: .media,
                id: id,
                ts: ts,
                seq: idx + 1,
                total: total,
                from: input.from,
                media: input.media,
                data: chunk.base64EncodedString()
            )
            if idx == 0 {
                env.filename = input.filename
                env.mime = input.mime
                env.totalBytes = totalBytes
                env.caption = input.caption
            }
            return env
        }
    }

    /// Reassembler for media envelopes — collects binary chunks into Data.
    public final class MediaReassembler {
        private struct Header {
            let from: MediaFrom
            let media: MediaKind
            let filename: String?
            let mime: String?
            let caption: String?
            let totalBytes: Int?
        }
        private struct State {
            var received: [Int: Envelope]
            var total: Int
            var firstTs: Int64
            var header: Header?
        }
        private var parts: [String: State] = [:]

        public init() {}

        public func gc(now: Int64 = Int64(Date().timeIntervalSince1970 * 1000), maxAgeMs: Int64 = 60_000) {
            for (id, st) in parts where now - st.firstTs > maxAgeMs {
                parts.removeValue(forKey: id)
            }
        }

        public func push(_ env: Envelope) -> ReassembledMedia? {
            guard env.kind == .media,
                  let total = env.total,
                  let seq = env.seq,
                  let from = env.from,
                  let media = env.media,
                  let data = env.data,
                  seq >= 1, seq <= total
            else { return nil }

            var st = parts[env.id] ?? State(received: [:], total: total, firstTs: env.ts, header: nil)
            if st.total != total { return nil }
            st.received[seq] = env
            if seq == 1 {
                st.header = Header(
                    from: from,
                    media: media,
                    filename: env.filename,
                    mime: env.mime,
                    caption: env.caption,
                    totalBytes: env.totalBytes
                )
            }
            parts[env.id] = st
            _ = data // silence unused warning

            if st.received.count < total { return nil }
            guard let header = st.header else { return nil }

            var combined = Data()
            for i in 1...total {
                guard let part = st.received[i],
                      let chunkB64 = part.data,
                      let chunkBytes = Data(base64Encoded: chunkB64)
                else { return nil }
                combined.append(chunkBytes)
            }
            if let expected = header.totalBytes, combined.count != expected {
                parts.removeValue(forKey: env.id)
                return nil
            }
            parts.removeValue(forKey: env.id)
            return ReassembledMedia(
                id: env.id,
                ts: env.ts,
                from: header.from,
                media: header.media,
                bytes: combined,
                filename: header.filename,
                mime: header.mime,
                caption: header.caption
            )
        }
    }
}

File: plugin/src/wire.ts
// Wire format for Pilot ↔ OpenClaw messages.
//
// One JSON object per UDP datagram. Keep it small — typical Pilot MTU is
// ~1200 bytes after framing, so the envelope must fit in one packet to avoid
// fragmentation. If text exceeds that, the sender splits into N envelopes
// sharing the same `id` and assembles via (`seq`, `total`).

export const WIRE_VERSION = 1 as const;

/** Maximum bytes for a single envelope's JSON encoding. */
export const MAX_ENVELOPE_BYTES = 1024;

/** Direction-agnostic media kinds. */
export type MediaKind = "image" | "file" | "audio";

/** A message from the user to the claw (mobile → claw). */
export type UserMessage = {
  v: typeof WIRE_VERSION;
  kind: "user";
  id: string;
  ts: number;
  text: string;
  seq?: number;
  total?: number;
  /**
   * Optional HMAC-SHA256 over canonicalized envelope (no `hmac` field), base64.
   * If the receiver has a shared secret configured, a valid HMAC bypasses the
   * pilot-address allowlist — letting any peer with the secret authenticate
   * without per-device address pinning.
   */
  hmac?: string;
};

/** A reply from the claw to the user (claw → mobile). */
export type AgentMessage = {
  v: typeof WIRE_VERSION;
  kind: "agent";
  id: string;
  ts: number;
  text: string;
  seq?: number;
  total?: number;
  hmac?: string;
};

/**
 * Media envelope — image / file / audio attachment.
 *
 * Binary payload travels as `data` (base64) and is chunked across N envelopes
 * sharing the same `id`. Metadata (filename, mime, totalBytes) lives ONLY in
 * the first envelope (seq=1) to keep subsequent chunks lean.
 *
 * On the wire, an image upload looks like:
 *   { v:1, kind:"media", media:"image", id:"abc", ts:..., seq:1, total:8,
 *     filename:"cat.jpg", mime:"image/jpeg", totalBytes:6543, caption:"my cat",
 *     data:"<base64>" }
 *   { v:1, kind:"media", id:"abc", ts:..., seq:2, total:8, data:"<base64>" }
 *   ...
 */
export type MediaMessage = {
  v: typeof WIRE_VERSION;
  kind: "media";
  /** Sender role — same semantics as user/agent for text. */
  from: "user" | "agent";
  /** What this is. Receivers may render differently per kind. */
  media: MediaKind;
  id: string;
  ts: number;
  /** Base64-encoded chunk payload. */
  data: string;
  seq: number;
  total: number;
  /** First-chunk-only fields. */
  filename?: string;
  mime?: string;
  totalBytes?: number;
  caption?: string;
  hmac?: string;
};

/** Delivery acknowledgement (either direction). */
export type Ack = {
  v: typeof WIRE_VERSION;
  kind: "ack";
  id: string;
  ts: number;
};

/** Error signal (either direction). */
export type WireError = {
  v: typeof WIRE_VERSION;
  kind: "error";
  id: string;
  ts: number;
  code: string;
  text: string;
};

export type Envelope = UserMessage | AgentMessage | MediaMessage | Ack | WireError;

export function encodeEnvelope(env: Envelope): Buffer {
  const json = JSON.stringify(env);
  const buf = Buffer.from(json, "utf8");
  if (buf.length > MAX_ENVELOPE_BYTES) {
    throw new Error(
      `envelope ${buf.length}B exceeds MAX_ENVELOPE_BYTES (${MAX_ENVELOPE_BYTES}); chunk before encoding`,
    );
  }
  return buf;
}

export function decodeEnvelope(buf: Buffer | Uint8Array): Envelope {
  const text = Buffer.isBuffer(buf)
    ? buf.toString("utf8")
    : Buffer.from(buf).toString("utf8");
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (e) {
    throw new Error(`envelope not JSON: ${(e as Error).message}`);
  }
  if (!isEnvelope(parsed)) {
    throw new Error(`envelope failed schema check: ${text.slice(0, 200)}`);
  }
  return parsed;
}

function isEnvelope(v: unknown): v is Envelope {
  if (!v || typeof v !== "object") return false;
  const o = v as Record<string, unknown>;
  if (o.v !== WIRE_VERSION) return false;
  if (typeof o.id !== "string" || o.id.length === 0) return false;
  if (typeof o.ts !== "number") return false;
  switch (o.kind) {
    case "user":
    case "agent":
      return typeof o.text === "string";
    case "media":
      return (
        typeof o.data === "string" &&
        typeof o.seq === "number" &&
        typeof o.total === "number" &&
        (o.media === "image" || o.media === "file" || o.media === "audio") &&
        (o.from === "user" || o.from === "agent")
      );
    case "ack":
      return true;
    case "error":
      return typeof o.code === "string" && typeof o.text === "string";
    default:
      return false;
  }
}

// ---------------------------------------------------------------------------
// HMAC signing / verification — shared-secret authorization layer.
// ---------------------------------------------------------------------------

/**
 * Canonicalize an envelope for HMAC. We strip the `hmac` field itself, then
 * JSON.stringify with sorted keys for stable output across encoders.
 */
function canonicalForHmac(env: Envelope): string {
  const copy = { ...(env as Record<string, unknown>) };
  delete copy["hmac"];
  const keys = Object.keys(copy).sort();
  const ordered: Record<string, unknown> = {};
  for (const k of keys) ordered[k] = copy[k];
  return JSON.stringify(ordered);
}

/**
 * Compute an HMAC-SHA256 over the canonicalized envelope (without `hmac`),
 * returning base64. Both sides MUST use the same canonicalization, so always
 * route through this function.
 */
export async function signEnvelope(env: Envelope, secret: string): Promise<string> {
  const { createHmac } = await import("node:crypto");
  const h = createHmac("sha256", secret);
  h.update(canonicalForHmac(env));
  return h.digest("base64");
}

/**
 * Verify an envelope's HMAC against a secret. Returns true iff the hmac field
 * is present AND matches the computed value (constant-time compare).
 */
export async function verifyEnvelope(env: Envelope, secret: string): Promise<boolean> {
  const provided = (env as { hmac?: string }).hmac;
  if (!provided) return false;
  const { createHmac, timingSafeEqual } = await import("node:crypto");
  const h = createHmac("sha256", secret);
  h.update(canonicalForHmac(env));
  const expected = h.digest();
  let providedBuf: Buffer;
  try {
    providedBuf = Buffer.from(provided, "base64");
  } catch {
    return false;
  }
  if (providedBuf.length !== expected.length) return false;
  return timingSafeEqual(providedBuf, expected);
}

/** Generate a short, monotonically-tending message id (24 base36 chars). */
export function newId(): string {
  const t = Date.now().toString(36).padStart(9, "0");
  const r = Math.floor(Math.random() * 36 ** 12)
    .toString(36)
    .padStart(12, "0");
  return `${t}${r}`;
}

/**
 * Split text into N user envelopes that each fit MAX_ENVELOPE_BYTES.
 * All chunks share the same `id`; assembly uses (`seq`, `total`).
 */
export function chunkUserText(text: string, id = newId()): UserMessage[] {
  return chunkText(text, id, "user") as UserMessage[];
}

export function chunkAgentText(text: string, id = newId()): AgentMessage[] {
  return chunkText(text, id, "agent") as AgentMessage[];
}

function chunkText(
  text: string,
  id: string,
  kind: "user" | "agent",
): Envelope[] {
  const ts = Date.now();
  // First try unsplit.
  const single = { v: WIRE_VERSION, kind, id, ts, text } as Envelope;
  if (Buffer.byteLength(JSON.stringify(single), "utf8") <= MAX_ENVELOPE_BYTES) {
    return [single];
  }
  // Binary search a chunk size that fits.
  const envOverhead = Buffer.byteLength(
    JSON.stringify({ v: WIRE_VERSION, kind, id, ts, text: "", seq: 1, total: 99 }),
    "utf8",
  );
  const room = MAX_ENVELOPE_BYTES - envOverhead - 8; // safety
  const slices: string[] = [];
  let i = 0;
  while (i < text.length) {
    let lo = 1;
    let hi = Math.min(text.length - i, room);
    let best = 1;
    while (lo <= hi) {
      const mid = (lo + hi) >> 1;
      const piece = text.slice(i, i + mid);
      if (Buffer.byteLength(piece, "utf8") <= room) {
        best = mid;
        lo = mid + 1;
      } else {
        hi = mid - 1;
      }
    }
    slices.push(text.slice(i, i + best));
    i += best;
  }
  return slices.map((piece, idx) => ({
    v: WIRE_VERSION,
    kind,
    id,
    ts,
    text: piece,
    seq: idx + 1,
    total: slices.length,
  })) as Envelope[];
}

// ---------------------------------------------------------------------------
// Media envelopes — chunk and reassemble binary payloads.
// ---------------------------------------------------------------------------

export type ChunkMediaInput = {
  from: "user" | "agent";
  media: MediaKind;
  bytes: Buffer | Uint8Array;
  filename?: string;
  mime?: string;
  caption?: string;
  id?: string;
  ts?: number;
};

/**
 * Split a binary payload into media envelopes that each encode within
 * MAX_ENVELOPE_BYTES. Header fields (filename, mime, totalBytes, caption)
 * are placed ONLY on seq=1 to keep subsequent chunks lean.
 */
export function chunkMedia(input: ChunkMediaInput): MediaMessage[] {
  const id = input.id ?? newId();
  const ts = input.ts ?? Date.now();
  const totalBytes = input.bytes.length;

  // Estimate first-chunk overhead with all metadata fields populated.
  const firstSample: MediaMessage = {
    v: WIRE_VERSION,
    kind: "media",
    from: input.from,
    media: input.media,
    id,
    ts,
    data: "",
    seq: 1,
    total: 999,
    filename: input.filename,
    mime: input.mime,
    totalBytes,
    caption: input.caption,
  };
  const firstOverhead = Buffer.byteLength(JSON.stringify(firstSample), "utf8");
  const firstDataRoom = MAX_ENVELOPE_BYTES - firstOverhead - 8;

  // Subsequent chunks omit the metadata fields → smaller overhead.
  const restSample: MediaMessage = {
    v: WIRE_VERSION,
    kind: "media",
    from: input.from,
    media: input.media,
    id,
    ts,
    data: "",
    seq: 999,
    total: 999,
  };
  const restOverhead = Buffer.byteLength(JSON.stringify(restSample), "utf8");
  const restDataRoom = MAX_ENVELOPE_BYTES - restOverhead - 8;

  // Each base64 char encodes 6 bits → 4 base64 chars per 3 source bytes.
  // Convert "room for base64 string" to "room for source bytes".
  const firstSrcBytes = Math.max(3, Math.floor(firstDataRoom / 4) * 3);
  const restSrcBytes = Math.max(3, Math.floor(restDataRoom / 4) * 3);

  const src = Buffer.isBuffer(input.bytes) ? input.bytes : Buffer.from(input.bytes);
  const chunks: Buffer[] = [];
  let offset = 0;

  // First chunk (smaller, due to metadata)
  const firstSlice = src.subarray(offset, Math.min(offset + firstSrcBytes, src.length));
  chunks.push(firstSlice);
  offset += firstSlice.length;

  while (offset < src.length) {
    const slice = src.subarray(offset, Math.min(offset + restSrcBytes, src.length));
    chunks.push(slice);
    offset += slice.length;
  }

  const total = chunks.length;
  return chunks.map((chunk, idx) => {
    const env: MediaMessage = {
      v: WIRE_VERSION,
      kind: "media",
      from: input.from,
      media: input.media,
      id,
      ts,
      data: chunk.toString("base64"),
      seq: idx + 1,
      total,
    };
    if (idx === 0) {
      if (input.filename !== undefined) env.filename = input.filename;
      if (input.mime !== undefined) env.mime = input.mime;
      env.totalBytes = totalBytes;
      if (input.caption !== undefined) env.caption = input.caption;
    }
    return env;
  });
}

export type ReassembledMedia = {
  id: string;
  ts: number;
  from: "user" | "agent";
  media: MediaKind;
  bytes: Buffer;
  filename?: string;
  mime?: string;
  caption?: string;
};

/** Reassembler for media envelopes — collects binary chunks into a Buffer. */
export class MediaReassembler {
  private parts = new Map<
    string,
    {
      received: Map<number, MediaMessage>;
      total: number;
      firstTs: number;
      header?: {
        from: "user" | "agent";
        media: MediaKind;
        filename?: string;
        mime?: string;
        caption?: string;
        totalBytes?: number;
      };
    }
  >();

  gc(now = Date.now(), maxAgeMs = 60_000): void {
    for (const [id, st] of this.parts) {
      if (now - st.firstTs > maxAgeMs) this.parts.delete(id);
    }
  }

  push(env: MediaMessage): ReassembledMedia | null {
    const { id, seq, total } = env;
    if (seq < 1 || seq > total) return null;
    let st = this.parts.get(id);
    if (!st) {
      st = { received: new Map(), total, firstTs: env.ts };
      this.parts.set(id, st);
    }
    if (st.total !== total) return null;

    st.received.set(seq, env);
    if (seq === 1) {
      st.header = {
        from: env.from,
        media: env.media,
        filename: env.filename,
        mime: env.mime,
        caption: env.caption,
        totalBytes: env.totalBytes,
      };
    }
    if (st.received.size < total) return null;
    if (!st.header) return null; // never got the metadata; drop

    const parts: Buffer[] = [];
    for (let i = 1; i <= total; i++) {
      const part = st.received.get(i);
      if (!part) return null;
      parts.push(Buffer.from(part.data, "base64"));
    }
    const bytes = Buffer.concat(parts);
    if (st.header.totalBytes !== undefined && bytes.length !== st.header.totalBytes) {
      // Size mismatch — drop and let the sender retry. This is the most likely
      // tampering signal at the wire layer.
      this.parts.delete(id);
      return null;
    }
    this.parts.delete(id);
    return {
      id,
      ts: env.ts,
      from: st.header.from,
      media: st.header.media,
      bytes,
      filename: st.header.filename,
      mime: st.header.mime,
      caption: st.header.caption,
    };
  }
}

/** State for reassembling chunked messages keyed by envelope id. */
export class Reassembler<T extends UserMessage | AgentMessage> {
  private parts = new Map<
    string,
    { received: Map<number, T>; total: number; firstTs: number }
  >();

  /** Drop reassembly state for ids older than `maxAgeMs`. */
  gc(now: number = Date.now(), maxAgeMs = 60_000): void {
    for (const [id, state] of this.parts) {
      if (now - state.firstTs > maxAgeMs) this.parts.delete(id);
    }
  }

  /**
   * Feed one envelope. If the full message has now arrived, returns it
   * (with `text` concatenated). Otherwise returns null.
   */
  push(env: T): T | null {
    if (env.total === undefined || env.total === 1) {
      return env;
    }
    const seq = env.seq;
    const total = env.total;
    if (seq === undefined || seq < 1 || seq > total) return null;
    let st = this.parts.get(env.id);
    if (!st) {
      st = { received: new Map(), total, firstTs: env.ts };
      this.parts.set(env.id, st);
    }
    if (st.total !== total) return null; // contradictory total → drop
    st.received.set(seq, env);
    if (st.received.size < total) return null;
    let combined = "";
    for (let i = 1; i <= total; i++) {
      const part = st.received.get(i);
      if (!part) return null;
      combined += part.text;
    }
    this.parts.delete(env.id);
    return { ...env, text: combined, seq: undefined, total: undefined } as T;
  }
}
